File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/RegisterManager.java
package io.treehopper.libraries;

import java.util.ArrayList;
import java.util.Arrays;

public class RegisterManager {
    protected ArrayList<Register> _registers = new ArrayList<>();
    protected IRegisterManagerAdapter _adapter;

    public RegisterManager(IRegisterManagerAdapter adapter) {
        _adapter = adapter;
    }

    public void read(Register register) {
        register.setBytes(_adapter.read(register.address, register.width));
    }

    public void readRange(Register start, Register end) {
        int count = (end.address + end.width) - start.address;
        byte[] bytes = _adapter.read(start.address, count);
        int i = 0;

        for (Register reg : _registers) {
            if (reg.address < start.address || reg.address > end.address)
                continue;

            reg.setBytes(Arrays.copyOfRange(bytes, i, reg.width + i));
            i += reg.width;
        }
    }

    public void write(Register register) {
        _adapter.write(register.address, register.getBytes());
    }

    public void writeRange(Register start, Register end) {
        ArrayList<Byte> bytes = new ArrayList<>();
        for (Register reg : _registers) {
            if (reg.address < start.address || reg.address > end.address)
                continue;

            byte[] readBytes = reg.getBytes();
            for (int i = 0; i < readBytes.length; i++)
                bytes.add(readBytes[i]);
        }

        byte[] bytesToWrite = new byte[bytes.size()];
        for (int i = 0; i < bytes.size(); i++) {
            bytesToWrite[i] = bytes.get(i);
        }

        _adapter.write(start.address, bytesToWrite);
    }
}

File: NET/API/Treehopper.Libraries/RegisterManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Treehopper.Libraries
{
    public class RegisterManager
    {
        protected List<Register> _registers = new List<Register>();
        private IRegisterManagerAdapter adapter;
        protected bool multiRegisterAccess;

        public RegisterManager(IRegisterManagerAdapter adapter, bool multiRegisterAccess)
        {
            this.adapter = adapter;
            this.multiRegisterAccess = multiRegisterAccess;
        }

        public async Task read(Register register)
        {
            var data = await adapter.read(register.Address, register.Width).ConfigureAwait(false);
            register.setBytes(data);
        }

        public async Task readRange(Register start, Register end)
        {
            if (multiRegisterAccess)
            {
                var count = (end.Address + end.Width) - start.Address;
                var bytes = await adapter.read(start.Address, count).ConfigureAwait(false);
                int i = 0;
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    reg.setBytes(bytes.Skip(i).Take(reg.Width).ToArray());
                    i += reg.Width;
                }
            }
            else
            {
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    await read(reg);
                }
            }
        }

        public Task write(Register register)
        {
            return adapter.write(register.Address, register.getBytes());
        }

        public async Task writeRange(Register start, Register end)
        {
            if (multiRegisterAccess)
            {
                List<byte> bytes = new List<byte>();
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    bytes.AddRange(reg.getBytes());
                }
                await adapter.write(start.Address, bytes.ToArray()).ConfigureAwait(false);
            }
            else
            {
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    await write(reg).ConfigureAwait(false);
                }
            }
        }
    }
}
